Support sophisticated backpacks

Nik77x opened the pull request:
Backpacks from sophisticated backpacks didn't render when they were equipped. Now backpacks render on the back, attached to the "Torso" bone so all animations work as expected.

All SB's rendering features also work correctly since it's the same code used by them for the normal humanoid character just with different transformations to the model to make it fit on the dragon's back.

The backpack transforms (position offset, rotation and scale) can be changed in the config, rendering can also be completely disabled.

Things I'm unsure about:
- I added another category to the client configuration called "Compat" to store all configs related to compatibility with other mods, I don't know if that's ok or maybe these things should be stored somewhere else.
- The RenderLayer class used to render the backpack is stored in the compat folder, not inside the client->render... since it's strictly made for compatibility reasons.

Quick video showcasing the change:


https://github.com/user-attachments/assets/5a92e0ee-f55d-4c98-8606-3a9be9eb84f8


(Youtube link for higher quality, github is limited to 10mb)
https://youtu.be/tDlFMIvoDLI

BlackAures1 commented:
This is a cool addition! Thanks. Did you consider that we now allow players to make custom models?

SiverDX commented:
https://github.com/DragonSurvivalTeam/DragonSurvival/blob/50a86ef8859ca9bfb5d03b746601d8f46b2d7b26/src/main/java/by/dragonsurvivalteam/dragonsurvival/registry/dragon/body/DragonBody.java#L53

https://github.com/DragonSurvivalTeam/DragonSurvival/blob/50a86ef8859ca9bfb5d03b746601d8f46b2d7b26/src/main/java/by/dragonsurvivalteam/dragonsurvival/client/models/DragonModel.java#L171-L181

you can customize the model / substitute your own when defining a body
and multiple can be valid at the same time, i.e. player x uses the default model and player y uses a custom one

Nik77x commented:
> player x uses the default model and player y uses a custom one

Oh, got it, the offsets are stored on the client so with other people's models that wouldn't match. I didn't think about this.

Then the position of the backpack has to be model-specific, or else the position has to be synced to the server.
There doesn't seem to be a good way of doing this...
I'll look into this.

seelderr commented:
I think a backpack bone is probably the way to go here; depending on the type of attachment to the bone, there could be further tuning to position whatever thing on the backpack bone that is being rendered. Ideally we don't use any configs here though, as any permutation of configs would not work with some custom body type. It needs to rely entirely on the bone so that any model is compatible with the rendering.

BlackAures1 commented:
I submitted a commit where I added a backpack bone - `BackpackBone`

Nik77x commented:
I've finally had time to work on this. Now The backpack follows the "BackpackBone". If it exists, the backpack will render to that position / rotation.
I also moved the bone to fit the backpack on the back.

I added backpack offset transforms to the body definitions and tried to reduce clipping with the wings as much as I could.
On The central and northern body types it looks pretty good, on eastern / southern I moved it back a bit, it still clips but I think it's acceptable, on the western it clips a lot when walking. I tried moving it back but it doesn't solve the problem, and it starts looking floaty because it's not on the torso anymore.

### Some screenshots:

- eastern / southern:
 
![Screenshot 2025-01-29 191836](https://github.com/user-attachments/assets/579326b8-f58d-4eeb-9730-8d6912bd9700)

- western clipping

![Screenshot 2025-01-29 193919](https://github.com/user-attachments/assets/446a6859-7677-472c-8c55-d382307c5f30)

An Idea I had to remove clipping, but I don't know if it looks good.

![Screenshot 2025-01-29 190657](https://github.com/user-attachments/assets/313c45dd-c22f-47b0-b023-860e54b4428d)

On the central and northern I didn't change anything besides making the backpack slightly smaller on the northern body.

(All these modifications can be made in the body definitions)

1 thing I need feedback on:
- I left the "renderBackpack" config option in the "Compat" category. I thought someone might want to turn backpack rendering off, so I left it. I don't know if I should remove it or move it somewhere else (I was thinking of removing the Compat category now since it's the only config option left and move it in Rendering).

BlackAures1 commented:
I think it's best to move it config to render. Thank you for your work! It's very cool!
Right now it has a conflict and I can't do a merge. Could you fix the conflict?
![image](https://github.com/user-attachments/assets/ecbe7fec-1322-4d74-81aa-670d425446d9)

Nik77x commented:
> Thank you for your work! It's very cool!
Thanks!
I moved the config and fixed the conflicts. Should be ready to merge now!

BlackAures1 commented:
If you have discord, I can give you a special role for helping with the mod. Email me.

Nik77x commented:
I actually do have a discord! and actually I am already in the server :)
> Email me.

Where can I find your email?

BlackAures1 commented:
BlackAures (blackaures) in discord
